Wrappers/Python/cil/optimisation/functions/SubsetSumFunction.py

```diff
@@ -303,39 +297,366 @@ def memory_reset(self):
+        Advances in neural information processing systems. 2013.
         """
 
-        output = x * 0.0
+        if not self.gradients_allocated:
+            self.memory_init(x)         
+            
+        # Select the next subset (uniformly at random by default). This generates self.subset_num
         self.next_subset()
-        self.functions[self.subset_num].gradient(x, out=output)
 
-        return output
+        # In the first iteration and if the snapshot image has just been updated, the current iterate and snapshot will be the same
+        # thus tmp2 = 0
+        if self.iter == 0 or self.iter % (self.update_frequency * self.num_subsets) == 0:
+            self.tmp2 = 0.0 * self.full_gradient
+        # Otherwise, compute the difference between the subset gradient at current iterate and at the snapshot in memory    
+        else:
+            # Compute new gradient for current subset, store in tmp1
+            # tmp1 = gradient F_{subset_num} (x)
+            self.functions[self.subset_num].gradient(x, out=self.tmp1) 
+
+            # Compute difference between current subset function gradient at current iterate (tmp1) and at snapshot, store in tmp2
+            # tmp2 = gradient F_{subset_num} (x) - gradient F_{subset_num} (snapshot)
+            self.tmp1.axpby(1., -1., self.functions[self.subset_num].gradient(self.snapshot), out=self.tmp2) 
+
+        # Compute the output: tmp2 + full_grad
+        if out is None:
+            ret = 0.0 * self.tmp2
+            self.tmp2.axpby(1., 1., self.full_gradient, out=ret)
+        else:
+            self.tmp2.axpby(1., 1., self.full_gradient, out=out)
+
+        self.iter += 1
+
+        # Check whether to update the full gradient and snapshot image
+        # This happens once every update_frequency epochs, i.e. if the current iterate is a multiple of update_frequency * num_subsets 
+        # Setting self.update_frequency to inf indicates that we never want to conduct the update
+        if np.isinf(self.update_frequency) == False and self.iter % (self.update_frequency * self.num_subsets) == 0: 
+            self.memory_update(x)
+
+        # Apply preconditioning
+        if self.precond is not None:
+            if out is None:
+                ret.multiply(self.precond(self.subset_num,x),out=ret)
+            else:
+                out.multiply(self.precond(self.subset_num,x),out=out)
+
+        if out is None:
+            return ret
+
+    def memory_init(self, x):
+        
+        """        
+            initialise the full gradient and the snapshot, and store in memory.
+        """
+        # Setting the gradient variables with the correct acquisition geometry
+        self.full_gradient = x * 0.0
+        self.tmp2 = x * 0.0
+        self.tmp1 = x * 0.0
+
+        # Initialise the gradient and the snapshot
+        self.memory_update(x)        
+        self.gradients_allocated = True 
```

Comment:
SVRG initialises not only the gradients but also the snapshot (so this name is maybe inaccurate). Kept now for consistency with SAG/A

Reply:
Variable (for all classes) could/should be renamed memory allocated. Really thats what it is.